Add std_normal_cdf, std_normal_lcdf, std_normal_lccdf and std_normal_rng

mcol opened the pull request:
## Summary

Adds implementations of `std_normal_cdf`, `std_normal_lcdf`, `std_normal_lccdf` and `std_normal_rng` obtained by simplifying code from the respective `normal_*` functions. I've also added the deprecated `_log` forms, but perhaps they are not needed now?

When these get merged, I'll add them to the docs. This fixes #839. 

## Tests

Added tests for all functions, again taken from existing tests for normal. In most cases, those were already testing for `mu = 0, sigma = 1`, so expected test values could be reused without changes. In the few remaining cases, they were recomputed using Wolfram Alpha.

## Side Effects

I had to modify `VectorRNGTestRig` and related classes so that they could handle the case of a RNG with no parameters. This was necessary as using `{}` for the good/bad values in parameter 1 (not present in `std_normal_rng` would still cause the `always_bad_values` to be appended: because of that, tests failed as no exceptions were thrown.

## Checklist

- [X] Math issue #839

- [X] Copyright holder: Marco Colombo

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [X] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

SteveBronder commented on stan/math/prim/prob/std_normal_cdf.hpp:
```diff
@@ -0,0 +1,82 @@
+    const T_partials_return scaled_y = y_dbl * INV_SQRT_TWO;
+    T_partials_return cdf_;
+    if (y_dbl < -37.5) {
+      cdf_ = 0.0;
```
underscore in name is usually reserved for member's of a class

mcol replied:
This pattern came from `normal_cdf`. I've fixed my use here.

bob-carpenter commented:
> underscore in name is usually reserved for member's of a class

I'd say always.  If other uses exist, they snuck by review.

mcol commented:
I've now added also `std_normal_rng`. For that to pass the RNG tests I had to modify `VectorRNGTestRig` and related classes so that they could handle the case of a RNG with no parameters.

As for variable names with a trailing underscore, there are a few other uses in the tree. I've fixed the use in `std_normal_cdf`, I'll fix the other files in a separate PR.

I think this is ready to review. Will this require changes in stanc3 in order to expose the new functions?

rok-cesnovar commented:
> Will this require changes in stanc3 in order to expose the new functions?

Yes, but that should be a simple PR. If you need any help with that, let me know. Once you have a list of signatures you want to add to the language it should be a simple task.

bbbales2 commented on test/unit/math/mix/prob/std_normal_cdf_log_test.cpp:
```diff
@@ -0,0 +1,20 @@
+TEST(mathMixScalFun, std_normal_lcdf_derivatives) {
+  auto f = [](const auto& y) { return stan::math::std_normal_lcdf(y); };
+
+  stan::test::expect_ad(f, -50.0);
```
Can you do these tests for std_normal_lpdf, std_normal_lcdf and std_normal_lccdf too?

Then we can get rid of some other tests and make sure we're staying up to date with the testing framework stuff.

mcol replied:
This test is for `std_normal_lcdf` already, but uses the existing (albeit confusing) naming convention of calling it `std_normal_cdf_log_test.cpp`. I'm adding mix tests for `_cdf`, `_lccdf` and `_lpdf`.

mcol replied:
In the `_cdf` test I'm seeing these failures that I'm not sure if I should worry about: for `stan::test::expect_ad(f, -25.0)` I get:
```
./test/unit/math/expect_near_rel.hpp:92: Failure
Value of: stan::math::is_nan(x1) && stan::math::is_nan(x2)
  Actual: false
Expected: true
expect_near_rel(4.7760521554589183e-134, -nan)
expect_near_rel; require items x1(i) = x2(i): grad_hessian() grad Hessian
```
while for `stan::test::expect_ad(f, -20.0)`:
```
./test/unit/math/expect_near_rel.hpp:96: Failure
Expected equality of these values:
  x1
    Which is: 2.20286e-85
  x2
    Which is: inf
expect_near_rel(2.2028583965234379e-85, inf)
```

bbbales2 replied:
The first doesn't really bother me that much cuz it's saying our numerics aren't perfect 25 standard deviations away from the mean on the non-log scale.

The second though... That's weird. The number on the left seems more right.

@bob-carpenter for at least one Q -- Is x1 or x2 the reference?

mcol replied:
I think x1 is finite differences, x2 is ad. The message is about `grad_hessian() grad Hessian` (sorry I cut off the last line). The test passes at -19.1 and starts failing at -19.2 with message `expect_near_rel(1.3097063068172137e-78, inf)`. Note that also `normal_cdf` is affected similarly, only it doesn't have such test.

bbbales2 replied:
Oooh, okay. Yeah if it's with 2nd derivatives that doesn't bother me. I thought it was values. That's fine.

bbbales2 commented on test/unit/math/mix/prob/std_normal_test.cpp:
```diff
@@ -0,0 +1,48 @@
+#include <gtest/gtest.h>
+#include <vector>
+
+std::vector<double> test_fun(double y) {
```
All the tests in this file can be replaced with something like:

```
auto f = [](const auto& y) { return stan::math::std_normal_lpdf(y); };
  
stan::test::expect_ad(f, -0.3);
stan::test::expect_ad(f, 0.0);
stan::test::expect_ad(f, 1.7);
```

mcol replied:
Cool!

bbbales2 commented on test/unit/math/fwd/prob/std_normal_test.cpp:
```diff
@@ -0,0 +1,12 @@
+#include <vector>
+
+TEST(ProbAgradDistributionsStdNormal, fwd) {
+  using stan::math::fvar;
```
These tests can be deleted if the expect_ad tests are added.

mcol replied:
Done.

bbbales2 commented on test/unit/math/rev/prob/std_normal_ccdf_log_test.cpp:
```diff
@@ -0,0 +1,116 @@
+#include <vector>
+
+TEST(std_normal_lccdf, tail) {
+  using stan::math::std_normal_lccdf;
```
These should go in the prim tests.

mcol replied:
Done.

bbbales2 commented on test/unit/math/rev/prob/std_normal_ccdf_log_test.cpp:
```diff
@@ -0,0 +1,116 @@
+
+  stan::math::recover_memory();
+}
+
```
Adding expect_ad tests removes the need for everything else in this file.

bbbales2 commented on test/unit/math/rev/prob/std_normal_cdf_log_test.cpp:
```diff
@@ -0,0 +1,228 @@
+#include <string>
+#include <vector>
+
+TEST(std_normal_lcdf, tails) {
```
These tests look like prim tests.

mcol replied:
Yes!

bbbales2 commented on test/unit/math/rev/prob/std_normal_cdf_log_test.cpp:
```diff
@@ -0,0 +1,228 @@
+
+  stan::math::recover_memory();
+}
+
```
Everything below gets replaced by expect_ad

bbbales2 commented on test/unit/math/rev/prob/std_normal_cdf_test.cpp:
```diff
@@ -0,0 +1,149 @@
+#include <string>
+#include <vector>
+
+TEST(std_normal_cdf, tails) {
```
These should be prim tests.

bbbales2 commented on test/unit/math/rev/prob/std_normal_cdf_test.cpp:
```diff
@@ -0,0 +1,149 @@
+
+  stan::math::recover_memory();
+}
+
```
Everything below can be replaced by expect_ad.

mcol commented on stan/math/prim/prob/std_normal_ccdf_log.hpp:
```diff
@@ -0,0 +1,20 @@
+ * @deprecated use <code>std_normal_lccdf</code>
+ */
+template <typename T_y>
+inline return_type_t<T_y> std_normal_ccdf_log(const T_y& y) {
```
I asked the same question in the description of the PR... :) I've added them only because I was not sure if Stan would expect those, or if we are making promises as to how functions can be called by users... I'll be very happy to remove them from this PR if we could. @bob-carpenter?

bbbales2 replied:
I had a look at this and we should add the deprecated stuff.

When dealing with truncation, things like:

```
x ~ normal(y, 1) T[, 1];
```

The old compiler still references the *cdf_log functions.

The new compiler has a bug it looks like cause it's not generating functions at all: https://github.com/stan-dev/stanc3/issues/414

I guess since the deprecated-ness isn't resolved upstream, we leave these in.

mcol commented on stan/math/prim/prob/std_normal_rng.hpp:
```diff
@@ -0,0 +1,34 @@
+#ifndef STAN_MATH_PRIM_PROB_STD_NORMAL_RNG_HPP
+#define STAN_MATH_PRIM_PROB_STD_NORMAL_RNG_HPP
+
+#include <stan/math/prim/meta.hpp>
```
Done.

mcol commented on test/unit/math/prim/prob/std_normal_test.cpp:
```diff
@@ -0,0 +1,97 @@
+   *
+   * It does so for all possible combinations of calling arguments.
+   */
+  check_quantiles_real_real(StdNormalTestRig());
```
Thanks, I got the gist!

mcol commented:
Thanks @bbbales2 for the awesome review! I think I've addressed all your comments, but there are still two points that need to be resolved: what we should do with the deprecated functions (add them or not?) and if the two tests in `test/unit/math/mix/prob/std_normal_cdf_test.cpp` that I commented out are a real issue or not.

bob-carpenter commented:
What are x1 and x2?  In expect_near, the arguments are symmetric. By convention for googletest, we try to put the expected value first and the computed value second.

> On Jan 6, 2020, at 9:36 PM, Ben Bales <notifications@github.com> wrote:
> 
> @bbbales2 commented on this pull request.
> 
> In test/unit/math/mix/prob/std_normal_cdf_log_test.cpp:
> 
> > @@ -0,0 +1,20 @@
> +#include <test/unit/math/test_ad.hpp>
> +#include <limits>
> +
> +TEST(mathMixScalFun, std_normal_lcdf_derivatives) {
> +  auto f = [](const auto& y) { return stan::math::std_normal_lcdf(y); };
> +
> +  stan::test::expect_ad(f, -50.0);
> 
> The first doesn't really bother me that much cuz it's saying our numerics aren't perfect 25 standard deviations away from the mean on the non-log scale.
> 
> The second though... That's weird. The number on the left seems more right.
> 
> @bob-carpenter for at least one Q -- Is x1 or x2 the reference?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>
